### PR TITLE
Add Publish Grading Feature

### PR DESCRIPTION
**Pull request by emptygx**
**To be merged with Frontend [#2856](https://github.com/source-academy/frontend/pull/2856)**

For implementation decisions, see Frontend PR.

## Changes
- Assessment Table: Add new is_grading_published field which determines if students can see their grades.
- Assessment Config Table: Add new is_grading_auto_published field which determines if grades are auto published for auto graded assessments.
- Backend changes xp/grading details, e.g. grader, autograding_status, back to default if is_grading_published is false. This prevents students from checking response from server to see their grades even though the Frontend hides it. 
- Notifications Types: Removed autograded and graded types. Added published and unpublished grading types.
- Notifications now created when grades are published/unpublished
- Implement relevent functions for publishing/unpublishing grades
- Admins are able to publish/unpublish all graded submissions
- Autograder can autopublish grades when all answers have a `success` autograding_status
- Updated seed to accommodate changes

### CODE REVIEW

**Comment by YaleChen299** [comment]
> > Hi! Please fix the format and CI and tag me again when the PR is ready. Thank you!
> 
> @YaleChen299 Hello! Sorry, I have fixed the formatting. Thank you!

No worries! Can you check the CI again? I will approve the CI run when I see then.

**Review by YaleChen299 — DISMISSED** [review]
Hi @emptygx, I have fixed the formatting and credo issues. However, a few tests are not passing. I suspect that the behaviour of `user_total_xp` is affected by the behaviour of published/unpublished grades. 

What is the expected behaviour for the published/unpublished grades? Is it that users will not see their grades if unpublished? 

Could you look into these test cases and fix them? Please document the correct behaviours in terms of the test cases, i.e. for any behaviours that may be affected by the publish grading feature, create at least one positive and one negative test case. 

Thank you so much for wrapping up this properly! I think this feature would make the grading activities much more user-friendly!

File: lib/cadet/accounts/notification_type.ex
Comment (YaleChen299) [inline lib/cadet/accounts/notification_type.ex:15]: let's keep the atom as snake case, as camel case is used for module names which are under the hood atoms.

File: lib/cadet/assessments/assessments.ex
Comment (YaleChen299) [reply lib/cadet/assessments/assessments.ex:256]: another thing you can do is just query the answers where the `a.is_grading_published` is true since we are only aggregating here.

File: lib/cadet/assessments/assessments.ex
Comment (YaleChen299) [inline lib/cadet/assessments/assessments.ex:884]: I think there is no need to use multi and transaction here since there is only a single update. May II ask what was the reason for this?

File: lib/cadet_web/admin_controllers/admin_assets_controller.ex
Comment (YaleChen299) [inline lib/cadet_web/admin_controllers/admin_assets_controller.ex:25]: Hi, what is the reason for this change? This should be unrelated to this PR?

File: lib/cadet/devices/devices.ex
Comment (YaleChen299) [inline lib/cadet/devices/devices.ex:211]: similarly here, this is unrelated to this PR right? Is it cuz it is failing for some test?

Comment (emptygx) [reply lib/cadet/devices/devices.ex:211]: Hi, yes it was failing the formatting checks and was added when I ran mix format

File: lib/cadet_web/admin_controllers/admin_assets_controller.ex
Comment (emptygx) [reply lib/cadet_web/admin_controllers/admin_assets_controller.ex:25]: Same as the other one it was auto-corrected while formatting

File: lib/cadet_web/admin_controllers/admin_grading_controller.ex
Comment (YaleChen299) [inline lib/cadet_web/admin_controllers/admin_grading_controller.ex:85]: hi please add tests for the controllers thanks!

**Comment by YaleChen299** [comment]
> (part of frontend issue #2522)
> 
> Currently, the grading states design and business logic cause ambiguity with resubmission of assessments being assumed to be graded immediately, and with grades being displayed when they should not.
> 
> This feature aims to solve this issue by introducing a new grading state `published` and a new backend boolean `is_grading_published`, allowing explicit publishing of student's grades on the frontend and preventing the fetching of grades if they have yet to be published, hence creating a better logic flow.
> 
> This pull request contains changes to be made to the backend to enable this new feature.

Re-thinking about the problem that we are trying to solve, I think we need to auto un-publish the grades when we unsubmit.
The main point of publish grading is to solve resubmitted submissions shown as `graded` to graders so that grader will not know if the student is graded after resubmission.
This logic is a bit different then whether the grade is publish to the student or not. Please correct me if i am wrong! Thanks!

**Comment by emptygx** [comment]
> Re-thinking about the problem that we are trying to solve, I think we need to auto un-publish the grades when we unsubmit. The main point of publish grading is to solve resubmitted submissions shown as `graded` to graders so that grader will not know if the student is graded after resubmission. This logic is a bit different then whether the grade is publish to the student or not. Please correct me if i am wrong! Thanks!

Yes, I agree that should be part of the feature, which has already been implemented as part of the logic on the front end, which auto-unpublishes grades whenever an assessment is unsubmitted. Sorry for not mentioning that in the PR description, will make sure to add it into future documentation! Should I add it to the PR description as well?

File: lib/cadet/assessments/assessments.ex
Comment (emptygx) [reply lib/cadet/assessments/assessments.ex:256]: I found this to be the easier and more straightforward way to filter the data, but if maintainability may be an issue, then doing it at the controller level might be better like you said. I am not too sure how to go about that though and may need some guidance!

Comment (YaleChen299) [reply lib/cadet/assessments/assessments.ex:256]: I agree that making this filtering at the controller level is a bit more troublesome. I think what we can do is at least separate this filtering from the main subquery selection. If you can make it into a function to be piped through with sematic name like `filter_published_xp`, I will be happy to accept that.

**Comment by GabrielCWT** [comment]
TODO:

Implement test for is_grading_published after this and #1088 are merged

File: test/cadet_web/admin_controllers/admin_grading_controller_test.exs
Comment (josh1248) [inline test/cadet_web/admin_controllers/admin_grading_controller_test.exs:1]: what a legend, frontend definitely needs someone who cares about tests 🌚

File: lib/cadet/assessments/assessments.ex
Comment (RichDom2185) [inline lib/cadet/assessments/assessments.ex:1420]: Typo: Incorrecr error message for unpublish all.

File: lib/cadet/assessments/assessments.ex
Comment (RichDom2185) [inline lib/cadet/assessments/assessments.ex:1350]: Correct me if I'm wrong, but since these are subqueries, they'll run first in entirety before any downstream filtering.

So in order to avoid the ~30s query from the joining of the large table, should we filter by course ID here first?

**Review by RichDom2185 — COMMENTED** [review]
Hi, thanks for working on this! Just some comments below, thanks!

File: lib/cadet/assessments/assessments.ex
Comment (GabrielCWT) [reply lib/cadet/assessments/assessments.ex:1342]: I managed to filter the Question table subquery but the Answer table subquery requires quite a few joins to obtain the course_id to filter by. I was wondering if it is worth the multiple joins since I'm not too sure about costs in SQL.

Joins:
Answer -> Question -> Assessment (course_id)

I also thought of combining the queries together since I already use Question -> Assessment but you cant use aggregate functions e.g. filter and count

**Review by RichDom2185 — APPROVED** [review]
LGTM! Let's merge this first to not block the other team's PRs. As to your question, let's discuss it after this is merged.

Thanks for the work!

**Pull request merged**
